### DISCUSSION TITLE
Expose all members of a struct to avoid maintainenace of two files for large projects

### DISCUSSION THREAD

**Discussion by hofmannu** [wjakob/nanobind#324] [Ideas]
I am currently using `nanobind` for a large project where the settings for many processing functions are stored in structs. For each of those option structs I want to now generate a binding for Python (used by application developers). Nevertheless, at the moment I have the problem that I need to expose each property individually. Could we add a command such as `def_prop_rw_all` that automatically generates the bindings to all the public members of the corresponding struct or class? I could not find a way so far to automate this process. Problematically like this I need to maintain the code for the properties in two places.

I did a search for similar questions but could not find anything so far. I also hope this is the right place to ask this question.

**Comment by wjakob** [comment]
nanobind can't do that as of yet since compile-time reflection is not part of the C++ standard. (I think it's been discussed by the C++ standardization committee but it's safe to say that this is not going to show up in compilers for years).

What you need is a binding generator -- something that parses a header file and dumps out the `.def_prop_rw_all` commands. People have built such things for pybind11, you can probably find some when googling for "pybind11 binding generator" or similar. None exists for nanobind as of yet, to my knowledge (though it's probably easy to modify since the binding syntax is so similar).
